### PR TITLE
`MeanAveragePrecision` box and segm at same time

### LINKED ISSUE #1905: Extract the mean average precision of both bbox and segm simultaneously in one run
## 🚀 Feature

In the current [implementaion](https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/detection/mean_ap.py#L236), one has to initialize MeanAveragePrecision() twice to get bbox and segmentation results. The implementation should support getting both bbox and segmentation metrics with a single initialization.

### Motivation

The pytorch [evaluation](https://github.com/pytorch/vision/blob/main/references/detection/coco_eval.py#L20) has the possibility to get both bbox and segmentation mAP in one run. This implementation has been out for a while now and I believe we should have a similar behaviour!

### Pitch

Rather than taking a [Literal](https://github.com/Lightning-AI/torchmetrics/blob/master/src/torchmetrics/detection/mean_ap.py#L236) of either a bbox or segm, the input argument `iou_types` (instead of `iou_type`) should take a list instead. The returned value should Dict but with first key being the iou_type.

Please let me know what you think!

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1905 

Make it possible to extract mean average precision for both box and segmentation input at the same time. This is a larger refactor of the internals of the map metric, but it should make it easier to add support for keypoint detection (https://github.com/Lightning-AI/torchmetrics/issues/1361) in the future.

I am fairly confident that this does not break backwards compatibility in any way as all original tests are passing without any changes.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--1928.org.readthedocs.build/en/1928/

### CODE REVIEW

**Comment by vsuryamurthy** [comment]
Perfect, thanks for this change!

File: src/torchmetrics/detection/helpers.py [hunk 1]
```diff
@@ -11,20 +11,26 @@
...
+    if not isinstance(iou_type, list):
+        iou_type = [iou_type]
+
+    item_val_name = []
+    if "bbox" in iou_type:
+        item_val_name.append("boxes")
+    if "segm" in iou_type:
+        item_val_name.append("masks")
+    if any(i not in ["bbox", "segm"] for i in iou_type):
         raise Exception(f"IOU type {iou_type} is not supported")
```
Comment (vsuryamurthy) [inline src/torchmetrics/detection/helpers.py:34]:
```suggestion
    if any(i not in ["bbox", "segm"] for i in iou_type):
        raise Exception(f"IOU type {iou_type} is not supported")
    item_val_name = []
    if "bbox" in iou_type:
        item_val_name.append("boxes")
    if "segm" in iou_type:
        item_val_name.append("masks")
```

Doesnt matter much but it would be best to first check if there are incorrect iou type and then do the rest.

File: src/torchmetrics/detection/mean_ap.py [hunk 2]
```diff
@@ -327,15 +387,20 @@ def update(self, preds: List[Dict[str, Tensor]], target: List[Dict[str, Tensor]]
         _input_validator(preds, target, iou_type=self.iou_type)
```
Comment (vsuryamurthy) [inline src/torchmetrics/detection/mean_ap.py:387]: Since the iou types are already validated in line 330: ```self.iou_type = _validate_iou_type_arg(iou_type)```, the ```_input_validator``` fn contains redundant validation of iou types. One validation check can be potentially removed.

File: src/torchmetrics/detection/helpers.py [hunk 3]
```diff
@@ -36,42 +42,57 @@ def _input_validator(
...
     for i, item in enumerate(targets):
-        if item[item_val_name].size(0) != item["labels"].size(0):
-            raise ValueError(
-                f"Input {item_val_name} and labels of sample {i} in targets have a"
-                f" different length (expected {item[item_val_name].size(0)} labels, got {item['labels'].size(0)})"
-            )
+        for ivn in item_val_name:
+            if item[ivn].size(0) != item["labels"].size(0):
+                raise ValueError(
+                    f"Input {ivn} and labels of sample {i} in targets have a"
```
Comment (Borda) [inline src/torchmetrics/detection/helpers.py:70]:
```suggestion
                    f"Input '{ivn}' and labels of sample {i} in targets have a"
```

File: src/torchmetrics/detection/helpers.py [hunk 4]
```diff
@@ -36,42 +42,57 @@ def _input_validator(
...
-        if not (item[item_val_name].size(0) == item["labels"].size(0) == item["scores"].size(0)):
-            raise ValueError(
-                f"Input {item_val_name}, labels and scores of sample {i} in predictions have a"
-                f" different length (expected {item[item_val_name].size(0)} labels and scores,"
-                f" got {item['labels'].size(0)} labels and {item['scores'].size(0)})"
-            )
+        for ivn in item_val_name:
+            if not (item[ivn].size(0) == item["labels"].size(0) == item["scores"].size(0)):
+                raise ValueError(
+                    f"Input {ivn}, labels and scores of sample {i} in predictions have a"
```
Comment (Borda) [inline src/torchmetrics/detection/helpers.py:77]:
```suggestion
                    f"Input '{ivn}', labels and scores of sample {i} in predictions have a"
```

File: src/torchmetrics/detection/helpers.py [hunk 5]
```diff
@@ -36,42 +42,57 @@ def _input_validator(
...
 def _fix_empty_tensors(boxes: Tensor) -> Tensor:
     """Empty tensors can cause problems in DDP mode, this methods corrects them."""
     if boxes.numel() == 0 and boxes.ndim == 1:
         return boxes.unsqueeze(0)
     return boxes
+
+
+def _validate_iou_type_arg(iou_type: Union[Literal["bbox", "segm"], List[str]] = "bbox") -> List[str]:
+    allowed_iou_types = ("segm", "bbox")
+    if not isinstance(iou_type, list):
```
Comment (Borda) [inline src/torchmetrics/detection/helpers.py:92]:
```suggestion
    if isinstance(iou_type, str):
```

File: src/torchmetrics/detection/mean_ap.py [hunk 6]
```diff
@@ -155,8 +155,8 @@ class MeanAveragePrecision(Metric):
                   width and height.
 
         iou_type:
-            Type of input (either masks or bounding-boxes) used for computing IOU.
-            Supported IOU types are ``["bbox", "segm"]``. If using ``"segm"``, masks should be provided in input.
+            Type of input (either masks or bounding-boxes) used for computing IOU. Supported IOU types are
+            ``"bbox"`` or ``"segm"`` or both as a list.
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:159]:
```suggestion
            ``"bbox"`` or ``"segm"`` or both as a tuple.
```
tuple would be safer with respect to some defaults

File: src/torchmetrics/detection/mean_ap.py [hunk 7]
```diff
@@ -350,68 +414,97 @@ def compute(self) -> dict:
...
-            map_per_class_values = torch.tensor([-1], dtype=torch.float32)
-            mar_100_per_class_values = torch.tensor([-1], dtype=torch.float32)
+            for i_type in self.iou_type:
+                if len(self.iou_type) > 1:
+                    # the area calculation is different for bbox and segm and therefore to get the small, medium and
+                    # large values correct we need to dynamically change the area attribute of the annotations
+                    for anno in coco_preds.dataset["annotations"]:
+                        anno["area"] = anno[f"area_{i_type}"]
+
+                self.coco_eval = COCOeval(coco_target, coco_preds, iouType=i_type)
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:442]: just thinking about wrapping this long boty to i private function/method

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@SkafteNicki, mind checking the collision withe the prefix PR? :flamingo:
